website: Add data consistency error docs for "planned value does not match config value" (#1232)
hashicorp/terraform-plugin-sdk: website/docs/plugin/sdkv2/resources/data-consistency-errors.mdx

An example root cause of this issue could be from API normalization, such as a JSON string being returned from an API and stored in state with differing whitespace then what was originally in config.

#### SDKv2 Example

Here is an example of an SDKv2 resource schema and terraform config that simulates this data consistency error:

```go
func thingResource() *schema.Resource {
	return &schema.Resource{
		// ...
		Schema: map[string]*schema.Schema{
			"word": {
				Type:     schema.TypeString,
				Optional: true,
				Computed: true,
				StateFunc: func(word interface{}) string {
                    // This simulates an API returning the 'word' attribute as all uppercase,
                    // which is stored to state even if it doesn't match the config or prior value.
					return strings.ToUpper(word.(string))
				},
			},
		},
	}
}
```

```hcl
resource "examplecloud_thing" "this" {
  word = "value"
}
```

> A warning log will be produced and the resulting state after applying a new resource would look like:
> ```json
> {
>   //...
>   "resources": [
>     {
>       "mode": "managed",
>       "type": "examplecloud_thing",
>       "name": "this",
>       "instances": [
>         {
>           "attributes": {
>             // This is invalid! Config has this stored as "value"
>             "word": "VALUE"
>           },
>         }
>       ]
>     }
>   ],
> }
> ```

Reviewer: I'm not sure folks will be used to the seeing the JSON state, so maybe best to just say what will happen:

```suggestion
A warning log will be produced and the resulting state after applying a new resource it will store the attribute state as `VALUE` instead of `value`.
```

Author: Good point -> ⚡

#### SDKv2 Example

Here is an example of an SDKv2 resource schema and terraform config that simulates this data consistency error:

```go
func thingResource() *schema.Resource {
	return &schema.Resource{
		// ...
		Schema: map[string]*schema.Schema{
			"word": {
				Type:     schema.TypeString,
				Optional: true,
				Computed: true,
				StateFunc: func(word interface{}) string {
                    // This simulates an API returning the 'word' attribute as all uppercase,
                    // which is stored to state even if it doesn't match the config or prior value.
					return strings.ToUpper(word.(string))
				},
			},
		},
	}
}
```

```hcl
resource "examplecloud_thing" "this" {
  word = "value"
}
```

A warning log will be produced and the resulting state after applying a new resource would look like:
```json
{
  //...
  "resources": [
    {
      "mode": "managed",
      "type": "examplecloud_thing",
      "name": "this",
      "instances": [
        {
          "attributes": {
            // This is invalid! Config has this stored as "value"
            "word": "VALUE"
          },
        }
      ]
    }
  ],
}
```

> #### Migrating to Plugin Framework

Author: I opened https://github.com/hashicorp/terraform-plugin-framework/pull/831 and moved the code over in 016f347

```text
examplecloud_thing.this: Modifying...
╷
│ Error: Provider produced inconsistent result after apply
│ 
│ When applying changes to examplecloud_thing.this, provider "provider[\"TYPE\"]" produced an unexpected
│ new value: .word: was cty.StringVal("value"), but now cty.StringVal("VALUE").
│ 
│ This is a bug in the provider, which should be reported in the provider's own issue tracker.
```

#### Recommended Solution
To solve this issue, the provider code must preserve the config value or prior state value when producing the new state. The recommended way to implement this logic is by creating a [custom type](/terraform/plugin/framework/handling-data/types/custom) with [semantic equality logic](/terraform/plugin/framework/handling-data/types/custom#semantic-equality). A custom type can be shared across multiple resource attributes and will ensure that the semantic equality logic is invoked during the `Read`, `Create`, and `Update` methods respectively.

For the above example, the semantic equality implementation below would resolve the resource drift and error:
```go
type CaseInsensitive struct {
	basetypes.StringValue
}

> // ... custom value type implementation

Reviewer 2: Not sure if it's worth adding further detail about what would need to be implemented for a custom value type, or perhaps linking to the relevant docs?

Author: Good suggestion. I'm going to move this section over to the migration guide in the framework repo, but I'll add a `<Note>` above this code example that says this isn't a full example and links to the custom value documentation 👍🏻

Author: Added suggestion in new PR, see https://github.com/hashicorp/terraform-plugin-framework/pull/831/files#diff-671bcafb4ced355e43bc6d152c9bef463a5fc27fa42d2ec2eaa4b9409004d33aR245-R249
